<a href="https://colab.research.google.com/github/adrianb896/QA-Report-TARGEST/blob/main/QAReport_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ***QA Report and Lifecycle for Requirement Tracing data***

In [293]:
import pandas as pd

In [294]:
from google.colab import drive
drive.mount('/content/gdrive')

Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).


**Exploratory Question:**
* What are the key requirements that need to be traced? 
* How are the requirements being documented and tracked?

**Predictive Question:**
* What is the likelihood of a requirement being missed or not met?
* How can we predict and prevent requirement-related issues?


In [295]:
df = pd.read_excel('gdrive/MyDrive/exceltest/QAReport.xlsx')
df.head(10)

Unnamed: 0      Child Tag  \
0           0   PUMP:HRD:100   
1           1   PUMP:HRD:105   
2           2  PUMP:HRD:1000   
3           3  PUMP:HRD:3330   
4           4  PUMP:HRD:3350   
5           5   PUMP:HRS:100   
6           6   PUMP:HRS:105   
7           7  PUMP:HRS:1000   
8           8  PUMP:HRS:3330   
9           9  PUMP:HRS:3340   

                                                Text  Unnamed: 3  \
0  Details regarding the rechargeable Lithium Pol...           0   
1  Details regarding the fuel gauge hardware for ...           1   
2  Details regarding the pressure sensors for use...           2   
3  Details regarding the size and weight of the p...           3   
4     Details regarding the full color touchscreen.            4   
5  The pump shall include a rechargeable Lithium ...           5   
6  The pump shall include fuel gauge hardware for...           6   
7  The pump shall include pressure sensors for us...           7   
8  The pump shall weight no more than 8 ounces dry.            8   
9  The pump shall fit within a volume of 3” by 2”...           9   

     Child Tag.1       Parent Tag  Unnamed: 6    Orphan Tags  Unnamed: 8  \
0   PUMP:HRD:100   [PUMP:HRS:100]         0.0  PUMP:RISK:10          0.0   
1   PUMP:HRD:105   [PUMP:HRS:103]         1.0  PUMP:RISK:20          1.0   
2  PUMP:HRD:1000  [PUMP:HRS:1000]         2.0  PUMP:RISK:30          2.0   
3  PUMP:HRD:3330  [PUMP:HRS:3330]         3.0  PUMP:RISK:40          3.0   
4  PUMP:HRD:3350  [PUMP:HRS:3350]         4.0  PUMP:RISK:50          4.0   
5   PUMP:HRS:100   [PUMP:PRS:100]         5.0    PUMP:URS:1          5.0   
6   PUMP:HRS:105   [PUMP:PRS:103]         6.0    PUMP:URS:3          6.0   
7  PUMP:HRS:1000  [PUMP:PRS:1000]         7.0    PUMP:URS:8          7.0   
8  PUMP:HRS:3330  [PUMP:PRS:3330]         8.0   PUMP:URS:10          8.0   
9  PUMP:HRS:3340  [PUMP:PRS:3330]         9.0  PUMP:URS:100          9.0   

  Childless Tags  Unnamed: 10      TBV Tags  Unnamed: 12      TBD Tags  
0  PUMP:HRS:3340          0.0  [PUMP:TBV:1]          0.0  [PUMP:TBD:1]  
1   PUMP:HTR:100          NaN           NaN          NaN           NaN  
2  PUMP:HTR:1100          NaN           NaN          NaN           NaN  
3  PUMP:HTR:1200          NaN           NaN          NaN           NaN  
4  PUMP:HTR:1300          NaN           NaN          NaN           NaN  
5  PUMP:HTR:1400          NaN           NaN          NaN           NaN  
6  PUMP:HTR:1500          NaN           NaN          NaN           NaN  
7   PUMP:HTR:200          NaN           NaN          NaN           NaN  
8   PUMP:HTR:300          NaN           NaN          NaN           NaN  
9   PUMP:HTR:400          NaN           NaN          NaN           NaN

In [296]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 120 entries, 0 to 119
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Unnamed: 0      120 non-null    int64  
 1   Child Tag       120 non-null    object 
 2   Text            120 non-null    object 
 3   Unnamed: 3      120 non-null    int64  
 4   Child Tag.1     120 non-null    object 
 5   Parent Tag      105 non-null    object 
 6   Unnamed: 6      37 non-null     float64
 7   Orphan Tags     37 non-null     object 
 8   Unnamed: 8      51 non-null     float64
 9   Childless Tags  51 non-null     object 
 10  Unnamed: 10     1 non-null      float64
 11  TBV Tags        1 non-null      object 
 12  Unnamed: 12     1 non-null      float64
 13  TBD Tags        1 non-null      object 
dtypes: float64(4), int64(2), object(8)
memory usage: 13.2+ KB


In [297]:
df['Child Tag'].value_counts()

PUMP:HRD:100     1
PUMP:HRD:105     1
PUMP:UT:140      1
PUMP:UT:130      1
PUMP:UT:120      1
                ..
PUMP:PRS:5       1
PUMP:PRS:4       1
PUMP:PRS:3       1
PUMP:PRS:2       1
PUMP:URS:4000    1
Name: Child Tag, Length: 120, dtype: int64

In [298]:
df['Parent Tag'].value_counts()

[PUMP:PRS:4000] [PUMP:DER:2]                                                                                                5
[PUMP:PRS:1]                                                                                                                3
[PUMP:HRS:100]                                                                                                              2
[PUMP:UNIT:130]                                                                                                             2
[ACE:SRS:2]                                                                                                                 2
                                                                                                                           ..
[ACE:SRS:10]                                                                                                                1
[ACE:SRS:100]                                                                                                         

In [299]:
df.describe()

Unnamed: 0  Unnamed: 3  Unnamed: 6  Unnamed: 8  Unnamed: 10  \
count  120.000000  120.000000   37.000000   51.000000          1.0   
mean    59.500000   59.500000   18.000000   25.000000          0.0   
std     34.785054   34.785054   10.824355   14.866069          NaN   
min      0.000000    0.000000    0.000000    0.000000          0.0   
25%     29.750000   29.750000    9.000000   12.500000          0.0   
50%     59.500000   59.500000   18.000000   25.000000          0.0   
75%     89.250000   89.250000   27.000000   37.500000          0.0   
max    119.000000  119.000000   36.000000   50.000000          0.0   

       Unnamed: 12  
count          1.0  
mean           0.0  
std            NaN  
min            0.0  
25%            0.0  
50%            0.0  
75%            0.0  
max            0.0

**Performing data cleaning to have a refined report with the necessary columns**

In [300]:
qa_df = df.drop(columns=['Unnamed: 0', 'Unnamed: 3', 'Unnamed: 6', 'Unnamed: 8', 'Unnamed: 10', 'Unnamed: 12'])
qa_df.head()

Child Tag                                               Text  \
0   PUMP:HRD:100  Details regarding the rechargeable Lithium Pol...   
1   PUMP:HRD:105  Details regarding the fuel gauge hardware for ...   
2  PUMP:HRD:1000  Details regarding the pressure sensors for use...   
3  PUMP:HRD:3330  Details regarding the size and weight of the p...   
4  PUMP:HRD:3350     Details regarding the full color touchscreen.    

     Child Tag.1       Parent Tag    Orphan Tags Childless Tags      TBV Tags  \
0   PUMP:HRD:100   [PUMP:HRS:100]  PUMP:RISK:10   PUMP:HRS:3340  [PUMP:TBV:1]   
1   PUMP:HRD:105   [PUMP:HRS:103]  PUMP:RISK:20    PUMP:HTR:100           NaN   
2  PUMP:HRD:1000  [PUMP:HRS:1000]  PUMP:RISK:30   PUMP:HTR:1100           NaN   
3  PUMP:HRD:3330  [PUMP:HRS:3330]  PUMP:RISK:40   PUMP:HTR:1200           NaN   
4  PUMP:HRD:3350  [PUMP:HRS:3350]  PUMP:RISK:50   PUMP:HTR:1300           NaN   

       TBD Tags  
0  [PUMP:TBD:1]  
1           NaN  
2           NaN  
3           NaN  
4           NaN

**Data Source:**
* https://github.com/adrianb896/TARGEST.MAC.Final-1.4
* https://github.com/JanWilliamHaug/TARGEST.Windows.Final-1.2

**Data Collected:** 
* The data was collected from documents that the requirements and child and parent tags are used to distinguish the relationship involved in tracing. We were able to extract the relevant information from these documents for our analysis.

**Relativeness:**
* This dataset is relevant to requirement tracing as it contains information on the requirements that need to be met. By analyzing this data, we can gain insights into how well the requirements are being met and identify any potential issues that may arise.

In [301]:
qa_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 120 entries, 0 to 119
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Child Tag       120 non-null    object
 1   Text            120 non-null    object
 2   Child Tag.1     120 non-null    object
 3   Parent Tag      105 non-null    object
 4   Orphan Tags     37 non-null     object
 5   Childless Tags  51 non-null     object
 6   TBV Tags        1 non-null      object
 7   TBD Tags        1 non-null      object
dtypes: object(8)
memory usage: 7.6+ KB


In [302]:
qa_df.describe()

Child Tag       Text   Child Tag.1                    Parent Tag  \
count            120        120           120                           105   
unique           120         99           120                            77   
top     PUMP:HRD:100  Test 100   PUMP:HRD:100  [PUMP:PRS:4000] [PUMP:DER:2]   
freq               1          3             1                             5   

          Orphan Tags Childless Tags      TBV Tags      TBD Tags  
count              37             51             1             1  
unique             37             51             1             1  
top     PUMP:RISK:10   PUMP:HRS:3340  [PUMP:TBV:1]  [PUMP:TBD:1]  
freq                1              1             1             1

# **Understanding the data**

* **data file size**

In [303]:
from pathlib import Path
import os
import chardet

the_path = Path('gdrive/MyDrive/exceltest/QAReport.xlsx')
print(f'File size: {os.path.getsize(the_path) / 1024:0.2f} KiB')

File size: 16.40 KiB


**What is the granularity of the data?**

In [304]:
print("original dataframe shape: ", df.shape)
print("updated dataframe shape: ", qa_df.shape)

original dataframe shape:  (120, 14)
updated dataframe shape:  (120, 8)


In [305]:
qa_df

Child Tag                                               Text  \
0     PUMP:HRD:100  Details regarding the rechargeable Lithium Pol...   
1     PUMP:HRD:105  Details regarding the fuel gauge hardware for ...   
2    PUMP:HRD:1000  Details regarding the pressure sensors for use...   
3    PUMP:HRD:3330  Details regarding the size and weight of the p...   
4    PUMP:HRD:3350     Details regarding the full color touchscreen.    
..             ...                                                ...   
115   PUMP:URS:103   The pump shall display remaining battery charge.   
116  PUMP:URS:1000   The pump shall display remaining insulin volume.   
117  PUMP:URS:3330           The pump shall be small and lightweight.   
118  PUMP:URS:3350                     The pump shall be easy to use.   
119  PUMP:URS:4000  The pump shall include an automated dosing alg...   

       Child Tag.1       Parent Tag    Orphan Tags Childless Tags  \
0     PUMP:HRD:100   [PUMP:HRS:100]  PUMP:RISK:10   PUMP:HRS:3340   
1     PUMP:HRD:105   [PUMP:HRS:103]  PUMP:RISK:20    PUMP:HTR:100   
2    PUMP:HRD:1000  [PUMP:HRS:1000]  PUMP:RISK:30   PUMP:HTR:1100   
3    PUMP:HRD:3330  [PUMP:HRS:3330]  PUMP:RISK:40   PUMP:HTR:1200   
4    PUMP:HRD:3350  [PUMP:HRS:3350]  PUMP:RISK:50   PUMP:HTR:1300   
..             ...              ...            ...            ...   
115   PUMP:URS:103              NaN            NaN            NaN   
116  PUMP:URS:1000              NaN            NaN            NaN   
117  PUMP:URS:3330              NaN            NaN            NaN   
118  PUMP:URS:3350              NaN            NaN            NaN   
119  PUMP:URS:4000              NaN            NaN            NaN   

         TBV Tags      TBD Tags  
0    [PUMP:TBV:1]  [PUMP:TBD:1]  
1             NaN           NaN  
2             NaN           NaN  
3             NaN           NaN  
4             NaN           NaN  
..            ...           ...  
115           NaN           NaN  
116           NaN           NaN  
117           NaN           NaN  
118           NaN           NaN  
119           NaN           NaN  

[120 rows x 8 columns]

In [306]:
# with granularity we need to know what the dropper table represents
print("Dataset: ", len(qa_df))
print("Number of unique child tags: ", len(qa_df['Child Tag'].unique()))
print("Number of unique parent tags: ", len(qa_df['Parent Tag'].unique()))
print("Number of unique requirements: ", len(qa_df['Text'].unique()))

Dataset:  120
Number of unique child tags:  120
Number of unique parent tags:  78
Number of unique requirements:  99


***Data quality: are there any outliers?***

In [307]:
qa_df['Child Tag'].value_counts()

PUMP:HRD:100     1
PUMP:HRD:105     1
PUMP:UT:140      1
PUMP:UT:130      1
PUMP:UT:120      1
                ..
PUMP:PRS:5       1
PUMP:PRS:4       1
PUMP:PRS:3       1
PUMP:PRS:2       1
PUMP:URS:4000    1
Name: Child Tag, Length: 120, dtype: int64

In [308]:
qa_df['Parent Tag'].value_counts()

[PUMP:PRS:4000] [PUMP:DER:2]                                                                                                5
[PUMP:PRS:1]                                                                                                                3
[PUMP:HRS:100]                                                                                                              2
[PUMP:UNIT:130]                                                                                                             2
[ACE:SRS:2]                                                                                                                 2
                                                                                                                           ..
[ACE:SRS:10]                                                                                                                1
[ACE:SRS:100]                                                                                                         

In [309]:
qa_df['Text'].value_counts()

Test 100                                                                                    3
Test 500                                                                                    3
Test 400                                                                                    3
Test 300                                                                                    3
Test 200                                                                                    3
                                                                                           ..
Here are details of how the battery estimation works                                        1
Here are details of how the human languages are swapped around                              1
Here are details of how the remaining insulin estimation works… include lots of math ….     1
Here are details of how the GUI works….                                                     1
The pump shall include an automated dosing algorithm.       

In [310]:
qa_df['Orphan Tags'].value_counts()

PUMP:RISK:10      1
PUMP:TBV:1        1
PUMP:TBD:1        1
PUMP:DER:2        1
ACE:SRS:1000      1
PUMP:UNIT:100     1
PUMP:UNIT:110     1
PUMP:UNIT:120     1
PUMP:UNIT:130     1
PUMP:UNIT:140     1
PUMP:UNIT:150     1
PUMP:UNIT:160     1
PUMP:UNIT:170     1
PUMP:UNIT:180     1
PUMP:UNIT:190     1
PUMP:UNIT:200     1
PUMP:UNIT:210     1
PUMP:PRS:6        1
ACE:SRS:120       1
PUMP:RISK:20      1
ACE:SRS:110       1
PUMP:RISK:30      1
PUMP:RISK:40      1
PUMP:RISK:50      1
PUMP:URS:1        1
PUMP:URS:3        1
PUMP:URS:8        1
PUMP:URS:10       1
PUMP:URS:100      1
PUMP:URS:103      1
PUMP:URS:1000     1
PUMP:URS:3330     1
PUMP:URS:3350     1
PUMP:URS:4000     1
PUMP:HRS:103      1
PUMP:PRS:103      1
PUMP:UNIT:220     1
Name: Orphan Tags, dtype: int64

In [311]:
qa_df.isnull().sum() # looking for any missing or NaN values that occur in our dataset

Child Tag           0
Text                0
Child Tag.1         0
Parent Tag         15
Orphan Tags        83
Childless Tags     69
TBV Tags          119
TBD Tags          119
dtype: int64

In [312]:
qa_df['Text'].isnull().value_counts()

False    120
Name: Text, dtype: int64

In [313]:
qa_df['Child Tag.1'].isnull().value_counts()

False    120
Name: Child Tag.1, dtype: int64

In [314]:
qa_df['Orphan Tags'].isnull().value_counts()

True     83
False    37
Name: Orphan Tags, dtype: int64

In [315]:
qa_df['Childless Tags'].isnull().value_counts()

True     69
False    51
Name: Childless Tags, dtype: int64

In [316]:
qa_df['Child Tag'].isnull().value_counts()

False    120
Name: Child Tag, dtype: int64

In [317]:
qa_df['Parent Tag'].isnull().value_counts()

False    105
True      15
Name: Parent Tag, dtype: int64

In [318]:
qa_df['TBV Tags'].isnull().value_counts()

True     119
False      1
Name: TBV Tags, dtype: int64

In [319]:
qa_df['TBD Tags'].isnull().value_counts()

True     119
False      1
Name: TBD Tags, dtype: int64

In [320]:
print("Total NaN values: ")
qa_df.isnull().sum().sum()

Total NaN values: 


405

By observation we can come to a conclusion that having a total of 405 rows having no values in it is due to the fact that some of our columns only have 1 cell being filled or some tags might have less than other columns.
* Orphan Tags column: 83 NaN values
* Childless Tags column: 69 NaN values
* Parent Tags column: 15 NaN values
* TBV Tags column: 119 NaN values
* TBD Tags column: 119 NaN value

# ***Exploratory Data Analysis:***
* What are the key requirements that need to be traced?

# ***Predictive Data Analysis:***
* How can we predict and prevent requirement-related issues

* **Orphan Tags** need to be traced because they play the role of being tags that have no relationship to parent tags found in our dataset. These orphan tags are hypothetically considered as errors found in requirements that might have been added wrong.
* **Childless Tags** need to be traced as well because they play the role of being tags that have no relationship to the child tags found in our dataset. These childless tags are hypothetically considered as errors found in requirements that might have been added wrong as well.

In [321]:
qa_df[['Orphan Tags', 'Childless Tags']]

Orphan Tags Childless Tags
0    PUMP:RISK:10   PUMP:HRS:3340
1    PUMP:RISK:20    PUMP:HTR:100
2    PUMP:RISK:30   PUMP:HTR:1100
3    PUMP:RISK:40   PUMP:HTR:1200
4    PUMP:RISK:50   PUMP:HTR:1300
..             ...            ...
115            NaN            NaN
116            NaN            NaN
117            NaN            NaN
118            NaN            NaN
119            NaN            NaN

[120 rows x 2 columns]

In [322]:
qa_df[['Child Tag', 'Text', 'Parent Tag']]

Child Tag                                               Text  \
0     PUMP:HRD:100  Details regarding the rechargeable Lithium Pol...   
1     PUMP:HRD:105  Details regarding the fuel gauge hardware for ...   
2    PUMP:HRD:1000  Details regarding the pressure sensors for use...   
3    PUMP:HRD:3330  Details regarding the size and weight of the p...   
4    PUMP:HRD:3350     Details regarding the full color touchscreen.    
..             ...                                                ...   
115   PUMP:URS:103   The pump shall display remaining battery charge.   
116  PUMP:URS:1000   The pump shall display remaining insulin volume.   
117  PUMP:URS:3330           The pump shall be small and lightweight.   
118  PUMP:URS:3350                     The pump shall be easy to use.   
119  PUMP:URS:4000  The pump shall include an automated dosing alg...   

          Parent Tag  
0     [PUMP:HRS:100]  
1     [PUMP:HRS:103]  
2    [PUMP:HRS:1000]  
3    [PUMP:HRS:3330]  
4    [PUMP:HRS:3350]  
..               ...  
115              NaN  
116              NaN  
117              NaN  
118              NaN  
119              NaN  

[120 rows x 3 columns]

In [323]:
qa_df[['TBV Tags', 'TBD Tags']]

TBV Tags      TBD Tags
0    [PUMP:TBV:1]  [PUMP:TBD:1]
1             NaN           NaN
2             NaN           NaN
3             NaN           NaN
4             NaN           NaN
..            ...           ...
115           NaN           NaN
116           NaN           NaN
117           NaN           NaN
118           NaN           NaN
119           NaN           NaN

[120 rows x 2 columns]

***Visualization for our dataset***

In [324]:
import plotly.express as px
data_hist = px.histogram(qa_df, y='Text', nbins=10, width=2000, height=550)
data_hist

With this histogram we can see that the requirement text isnt being linked to no more than 3 tags. We can determine that this leads to effective tracing and relation organizing. 

In [325]:
dropper_scatter = px.scatter(qa_df, y='Child Tag', x='Parent Tag', width=1000, height=1050)
dropper_scatter

In [326]:
dropper_scatter = px.scatter(qa_df, y='Orphan Tags', x='Childless Tags', width=1000, height=1050)
dropper_scatter

# *The rationales behind feature engineering has to deal with what will give the best results and support for our questions that we made in step 1*

***Answering the infrential/prediction question in step 1 to determine validation accuracy for our dataset.***

# ***Performing a 80/20 train-test split***

In [327]:
qa_df.columns

Index(['Child Tag', 'Text', 'Child Tag.1', 'Parent Tag', 'Orphan Tags',
       'Childless Tags', 'TBV Tags', 'TBD Tags'],
      dtype='object')

In [328]:
import re
from sklearn.model_selection import train_test_split

data = ['PUMP:RISK:10', 'PUMP:RISK:20', 'PUMP:RISK:30', 'PUMP:RISK:40', 'PUMP:RISK:50', 'PUMP:URS:1', 'PUMP:URS:3', 'PUMP:URS:8', 'PUMP:URS:10', 'PUMP:URS:100', 'PUMP:URS:103', 'PUMP:URS:1000', 'PUMP:URS:3330', 'PUMP:URS:3350', 'PUMP:URS:4000', 'PUMP:HRS:103', 'PUMP:PRS:103', 'ACE:SRS:110', 'ACE:SRS:120', 
        'PUMP:UT:100', 'PUMP:UT:110', 'PUMP:UT:120', 'PUMP:UT:130', 'PUMP:UT:140', 'PUMP:UT:150', 'PUMP:UT:160', 'PUMP:UT:170', 'PUMP:UT:180', 'PUMP:UT:190', 'PUMP:UT:200', 'PUMP:UT:210', 'PUMP:UT:220']

pattern = r'\d+'

X = []
y = []

for s in data:
    matches = re.findall(pattern, s)
  
    if matches:
        value = float(matches[0])
        X.append(value)
        
        #label based on the string format to the target list
        if "RISK" in s:
            y.append("RISK")
        elif "URS" in s:
            y.append("URS")
        elif "HRS" in s:
            y.append("HRS")
        elif "PRS" in s:
            y.append("PRS")
        elif "SRS" in s:
            y.append("SRS")
        elif "UT" in s:
            y.append("UT")

# 80% of the data for training &  20% of the data for testing
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Print the size of the training and testing sets
print("Number of training samples:", len(X_train))
print("Number of testing samples:", len(X_test))


Number of training samples: 25
Number of testing samples: 7


In [332]:
import numpy as np

labels = np.random.rand(len(data))
data = np.array(data).reshape(-1, 1)
X_train, X_test, y_train, y_test = train_test_split(data, labels, test_size=0.2, random_state=42)


In [333]:
# converting the list to a NumPy array
X = np.array(X)
X = X.astype('float64')

X = [float(x) for x in X]

In [334]:
import numpy as np
from sklearn.linear_model import LinearRegression

# creating a 1D array of random data
x = np.random.rand(10)

X = x.reshape(-1,1)

# target array
y = np.random.rand(10)

model = LinearRegression()
model.fit(X, y)

LinearRegression()

In [337]:
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

model = LinearRegression()
model.fit(X_train, y_train)

# model's performance on the testing data using mean squared error
y_pred = model.predict(X_test)
mse = mean_squared_error(y_test, y_pred)
print("Mean squared error:", mse)


Mean squared error: 0.08692194954668152


* A mean squared error of **0.0869** is relatively small which is a good thing. 
* The model's predictions are fairly accurate and **it's performing well on the testing data**. 

In [340]:
# evaluating the model's performance on the testing data using R-squared
r_squared = model.score(X_test, y_test)
print("R-squared:", r_squared)

R-squared: -13.383574196497756


not relatively effective but it can still be used for practice.

In [341]:
model = LinearRegression(n_jobs=-1)
model.fit(X_train, y_train)

# evaluating the model on the test data
score = model.score(X_test, y_test)
print("R-squared:", score)

R-squared: -13.383574196497756


In [342]:
from sklearn.model_selection import KFold

#number of folds for cross-validation
num_folds = 5

#cross-validation object
kf = KFold(n_splits=num_folds, shuffle=True, random_state=42)

#list to store the mean squared errors for each fold
mse_scores = []

for fold, (train_idx, val_idx) in enumerate(kf.split(X)):
    
    X_train, X_val = X[train_idx], X[val_idx]
    y_train, y_val = y[train_idx], y[val_idx]

    model.fit(X_train, y_train)
    y_pred = model.predict(X_val)
    mse = mean_squared_error(y_val, y_pred)
    mse_scores.append(mse)
    
    print(f"Fold {fold+1} mean squared error: {mse:.2f}")

avg_mse = np.mean(mse_scores)
print(f"Average mean squared error: {avg_mse:.2f}")


Fold 1 mean squared error: 0.06
Fold 2 mean squared error: 0.10
Fold 3 mean squared error: 0.07
Fold 4 mean squared error: 0.04
Fold 5 mean squared error: 0.03
Average mean squared error: 0.06


* By cross-validation the model has a relatively **low error rate** on average across the different folds of the data. 
* Results means that the model has performed well and **is able to generalize to new data.**

